aten/src/ATen/native/cuda/ReduceOpsKernel.cu

```diff
@@ -222,6 +224,227 @@ void argmin_kernel_cuda(TensorIterator& iter) {
+          row_buf[offset + d] = out;
+        }
+        __syncthreads();
+      }
+
+      // Write back to output.
+      if (row < num_rows) {
+        if (col1 < row_size){
+          row_values[col1] = row_buf[threadIdx.x];
+          row_indices[col1] = row_idx_buf[threadIdx.x];
+        }
+        if (col2 < row_size) {
+          row_values[col2] = row_buf[num_threads_x + threadIdx.x];
+          row_indices[col2] = row_idx_buf[num_threads_x + threadIdx.x];
+        }
+      }
+      block_total = row_buf[2 * num_threads_x - 1];
+      block_idx_final = row_idx_buf[2 * num_threads_x - 1];
+      __syncthreads();
+    }
+  }
+}
+
+/* Perform an inclusive scan along an outer dimension of a tensor.
+ *
+ * - num_orows is the size of the flattened outer dimensions;
+ * - num_irows is the size of the flattened inner dimensions;
+ * - row_size is the size of the dimension along which to compute the variance;
+ *
+ * The dimensions to the outside and inside of the specified dimension are considered as flattened.
+ * Thread blocks with the same blockIdx.y process an "outer row" (i.e. an element of the flattened
+ * outer dimensions, which contains several "inner rows").
+ * Each thread processes a single inner row at a time.
+ */
+template<typename scalar_t, class BinaryFunction>
+__global__ void tensor_kernel_scan_outer_dim_with_indices(scalar_t *self_, scalar_t *values_, int64_t *indices_,
+                  int num_orows, int num_irows, int row_size, scalar_t init, BinaryFunction binary_op) {
+  for (int orow = blockIdx.x; orow < num_orows; orow += gridDim.x) {
+    for (int irow = blockIdx.y * blockDim.x + threadIdx.x; irow < num_irows; irow += gridDim.y * blockDim.x) {
+      scalar_t *self = self_ + orow * row_size * num_irows + irow;
+      scalar_t *values = values_ + orow * row_size * num_irows + irow;
+      int64_t *indices = indices_ + orow * row_size * num_irows + irow;
+      scalar_t out = init;
+      int64_t idx = 0;
+
+      for (int col = 0; col < row_size; ++col) {
+        out = binary_op(out, *self);
+        if(out == *self) {
+          idx = col;
+        }
+        *values = out;
+        *indices = idx;
+        self += num_irows;
+        values += num_irows;
+        indices += num_irows;
+      }
+    }
+  }
+}
+
+template<typename scalar_t, class BinaryFunction>
+__host__ void scan_outer_dim_with_indices(const Tensor& self, Tensor& values, Tensor& indices,
+                                       int dim, scalar_t init, BinaryFunction binary_op) {
+  int ndim = self.dim();
+  // Treat all outer dimensions (i.e. dim_ < dim) as one.
+  int num_orows = 1;
+  for (int dim_ = 0; dim_ < dim; dim_++) {
+    num_orows *= self.size(dim_);
+  }
+  int row_size = self.size(dim);
+  // Treat all inner dimensions (i.e. dim > dimension) as one.
+  int num_irows = 1;
+  for (int dim_ = dim + 1; dim_ < ndim; dim_++) {
+    num_irows *= self.size(dim_);
+  }
+
+  dim3 threads(std::min(512, num_irows));
+  int maxGridDim = 1024;
```

Comment:
You should not hardcode maxGridDim to arbitrary values, actual max dims can be queried using `at::cuda::getCurrentDeviceProperties()->maxGridSize;`, look in the code at how it's used